### PR TITLE
Support MCV based cardinality estimation for text columns

### PR DESCRIPTION
**Pull request by asubramanya**
Prior to this patch, the MCVs was not being used to compute the cardinality for
text based columns. This would cause errors in the cardinality and lead to a
sub optimal plan being picked. This patch enables ORCA to use the MCVs for
equal and not equal operators for text columns.

Co-authored-by: Abhijit Subramanya <asubramanya@pivotal.io>
Co-authored-by: Bhuvnesh Chaudhary <bchaudhary@pivotal.io>

### CODE REVIEW

File: data/dxl/minidump/SelectOnBpchar.mdp [hunk 1]
```diff
@@ -1330,7 +1330,7 @@
     <dxl:Plan Id="0" SpaceSize="1">
       <dxl:GatherMotion InputSegments="0,1" OutputSegments="-1">
         <dxl:Properties>
-          <dxl:Cost StartupCost="0" TotalCost="6457.783233" Rows="15116.850832" Width="125"/>
+          <dxl:Cost StartupCost="0" TotalCost="3689.988281" Rows="1.000000" Width="125"/>
```
Comment (chrishajas) [inline data/dxl/minidump/SelectOnBpchar.mdp]: Is this cardinality correct? The query from this appears to be something like 
`select * from lineitem where  L_SHIPINSTRUCT = 'some text';`

There are only 4 distinct values for this column, and each has a frequency of ~.25. Therefore, it seems like there should be closer to 15k rows, not 1.

Comment (asubramanya) [reply data/dxl/minidump/SelectOnBpchar.mdp]: It depends on what 'some_text' is. If it is one of those 4 distinct values then yes it should be ~15k. But, If you look at the constant in the query `AAAAFURFTElWRVIgSU4gUEVSU09O`, it does not match any of the 4 distinct values for the `L_SHIPINSTRUCT` column. So I think the estimate is more accurate in this case than before.

Comment (chrishajas) [reply data/dxl/minidump/SelectOnBpchar.mdp]: Also, this column is of type character (not varchar). Do we need to handle character too on the gpdb side so we don't hash the padding?

File: libnaucrates/src/base/IDatumStatisticsMappable.cpp [hunk 2]
```diff
@@ -54,11 +53,6 @@ IDatumStatisticsMappable::StatsAreEqual
 		return false;
 	}
 
-	if (is_binary_comparison)
-	{
-		return StatsEqualBinary(datum);
```
Comment (chrishajas) [inline libnaucrates/src/base/IDatumStatisticsMappable.cpp]: I don't see anywhere that we're calling `StatsEqualBinary()` or `StatsLessThanBinary()`. Do we want to remove those functions too?

Comment (asubramanya) [reply libnaucrates/src/base/IDatumStatisticsMappable.cpp]: Removed these functions.

File: server/src/unittest/dxl/statistics/CMCVTest.cpp [hunk 3]
```diff
@@ -178,12 +178,6 @@ CMCVTest::EresUnittest_MergeHistMCV()
 		"../data/dxl/statistics/Merge-Input-MCV-Numeric.xml",
 		"../data/dxl/statistics/Merge-Input-Histogram-Numeric.xml",
 		"../data/dxl/statistics/Merge-Output-Numeric.xml"
-		},
-
-		{
-		"../data/dxl/statistics/Merge-Input-MCV-BPChar.xml",
-		"../data/dxl/statistics/Merge-Input-Histogram-BPChar.xml",
-		"../data/dxl/statistics/Merge-Output-BPChar.xml"
 		}
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply server/src/unittest/dxl/statistics/CMCVTest.cpp]: I've added the tests in ICG.

File: libnaucrates/src/statistics/CHistogram.cpp [hunk 4]
```diff
@@ -1033,6 +1036,25 @@ CHistogram::CopyHistogram
 	return histogram_copy;
 }
 
+BOOL
+CHistogram::SupportsTextFilter
+(
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply libnaucrates/src/statistics/CHistogram.cpp]: Fixed.

File: libnaucrates/src/base/CDatumGenericGPDB.cpp [hunk 5]
```diff
@@ -334,14 +334,11 @@ CDatumGenericGPDB::IsDatumMappableToLINT() const
 BOOL
 CDatumGenericGPDB::SupportsBinaryComp
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply libnaucrates/src/base/CDatumGenericGPDB.cpp]: Removed this function.

File: libnaucrates/src/base/CDatumGenericGPDB.cpp [hunk 6]
```diff
@@ -323,27 +323,6 @@ CDatumGenericGPDB::IsDatumMappableToLINT() const
...
-CDatumGenericGPDB::SupportsBinaryComp
-		(
-				const IDatum *datum_other
-		)
-	const
-{
-	return ((MDId()->Equals(&CMDIdGPDB::m_mdid_bpchar)
-			|| MDId()->Equals(&CMDIdGPDB::m_mdid_varchar)
-			|| MDId()->Equals(&CMDIdGPDB::m_mdid_text))
-			&& (this->MDId()->Sysid().Equals(datum_other->MDId()->Sysid())));
```
Comment (bhuvnesh2703) [inline libnaucrates/src/base/CDatumGenericGPDB.cpp]: what is the purpose of the sysid comparision? do we need to consider it while we check for text related types?

File: libnaucrates/src/statistics/CHistogram.cpp [hunk 7]
```diff
@@ -592,17 +592,20 @@ CHistogram::IsValid
 		return false;
 	}
 
-	for (ULONG bucket_index = 1; bucket_index < m_histogram_buckets->Size(); bucket_index++)
+	if (!IsHistogramForTextRelatedTypes())
```
Comment (zellerh) [inline libnaucrates/src/statistics/CHistogram.cpp]: Just wondering how we could or should check validity for text histograms. Could we check that every bucket in a text histogram is a singleton?

File: libnaucrates/src/statistics/CStatsPredUtils.cpp [hunk 8]
```diff
@@ -294,6 +302,14 @@ CStatsPredUtils::GetPredStats
 	GPOS_ASSERT(NULL != scalar_const_op);
 
 	IDatum *datum = scalar_const_op->GetDatum();
+
+	BOOL is_text_related_type = IsTextRelatedType(datum->MDId()) && IsTextRelatedType(col_ref->RetrieveType()->MDId());
```
Comment (zellerh) [inline libnaucrates/src/statistics/CStatsPredUtils.cpp]: I wonder whether this "&&" should be an "||"? Is it possible that someone could declare a comparator that compares a text column with an integer constant? In that case (if it is even possible), we would also not support the histogram comparison? Maybe a more general question - could we come across a user-defined comparison operator ```scalar_cmp_op``` here?

File: libnaucrates/src/statistics/CHistogram.cpp (hunk 7 above)
Comment (asubramanya) [reply libnaucrates/src/statistics/CHistogram.cpp]: Yes. Thats probably a good check to add. I'll create a separate story to track that.

File: libnaucrates/src/statistics/CStatsPredUtils.cpp (hunk 8 above)
Comment (asubramanya) [reply libnaucrates/src/statistics/CStatsPredUtils.cpp]: We don't want to support user defined comparison operators for now. We will fall back to the old behavior in such cases.

Comment (zellerh) [reply libnaucrates/src/statistics/CStatsPredUtils.cpp]: Thanks for looking at this and for the explanation.

**Pull request merged**